src/Servers/Kestrel/Core/src/Internal/Http2/FlowControl/OutputFlowControl.cs

```diff
@@ -57,9 +72,9 @@ public bool TryUpdateWindow(int bytes)
         {
             if (_flow.TryUpdateWindow(bytes))
             {
-                while (_flow.Available > 0 && _awaitableQueue?.Count > 0)
+                while (_flow.Available > 0 && _awaitableQueueIndex >= 0)
                 {
-                    _awaitableQueue.Dequeue().Complete();
+                    Dequeue().TrySetResult(null);
```

Comment:
Changing this from FIFO to LIFO will lead to some *really unfair* behavior when multiple streams are waiting on a connection level WINDOW_UPDATE.

At the stream level (`_streamLevelFlowControl`), there's only ever at most one writer `ProcessDataWrites()` at a time waiting for a WINDOW_UPDATE for a given stream. In this case, `_awaitableQueue` always has a length of 1 or 0 meaning that FIFO vs LIFO doesn't matter. I originally considered using a different type that doesn't use a queue for `_streamLevelFlowControl`, but I found that optimization premature. Maybe it's time to reconsider that decision.

At the connection level (`_connectionLevelFlowControl`), where many streams can be waiting for the connect level WINDOW_UPDATE, FIFO vs LIFO is a big deal.

With the previous logic, when a connection level WINDOW_UPDATE is received, the stream that's been waiting the longest gets the first crack of using the available credit. The stream will use as much of the available credit as it can, but if there is more connection-level availability than data the longest-waiting stream had to write (`_flow.Available > 0`), the previous logic would move to the next longest-waiting stream and let that attempt to write, and so on and so forth. `Complete()` and now `TrySetResult(null)` run continuations inline, so `_flow.Available` gets updated with every iteration of the loop.

With the new logic, the stream that just used all all the connection-level availability the last time gets first crack to do the same again the next time Kestrel receives a connection level WINDOW_UPDATE. Ultimately, this could easily cause a single stream to dominate all the writes.

I didn't write a test to verify the FIFO behavior, since at first I considered a bit of an implementation detail, but now I think it's important we test that we round-robin through the streams when there are multiple streams waiting on connection level WINDOW_UPDATEs.

---

**Side note about the algorithm:** It would also be correct to dequeue all the awaitables and complete them all at once every time we receive a connection level WINDOW_UPDATE. Every awaiting stream technically has the right to use the connection level window space (if there isn't stream level window space available already, the stream would wait for a stream level WINDOW_UPDATE before waiting on a connection level WINDOW_UPDATE) and locks ensure the `_connectionLevelFlowControl` is kept synchronized, but that would lead to the [thundering herd problem](https://en.wikipedia.org/wiki/Thundering_herd_problem).

Reply:
For some reason I had in my mind that the queue was a stack.

I'll change it, and see about adding a test.